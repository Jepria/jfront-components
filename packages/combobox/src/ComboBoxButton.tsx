import React from "react"
import {
  ComboBoxButtonProps,
  StyledButton,
  StyledSpan,
  StyledSvg,
} from "./styles"

export const ComboBoxButton = ({ rotate, ...props }: ComboBoxButtonProps) => {
  return (
    <StyledButton rotate={rotate} {...props} type="button">
      <StyledSpan>
        <StyledSvg>
          <path d="M7 9l5 5 5-5z" />
        </StyledSvg>
      </StyledSpan>
    </StyledButton>
  )
}
