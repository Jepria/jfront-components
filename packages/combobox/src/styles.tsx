import styled from "styled-components"

export interface ItemProps {
  disabled?: boolean
  selected?: boolean
  hover?: boolean
}

export const Item = styled.div<ItemProps>`
  box-sizing: border-box;
  overflow: hidden;
  white-space: nowrap;
  height: ${(props) => props.theme.combobox.item.height};
  padding: ${(props) => props.theme.combobox.item.padding};
  margin: ${(props) => props.theme.combobox.item.margin};
  font-size: ${(props) => props.theme.fontSize.medium};
  font-family: ${(props) => props.theme.fontFamily};
  font-size: ${(props) => props.theme.fontSize.medium};
  font-family: ${(props) => props.theme.fontFamily};
  text-align: left;
  border: ${(props) =>
    `${props.theme.combobox.item.borderWidth} 
    ${props.theme.combobox.item.borderStyle} 
    ${props.theme.combobox.item.borderColor}`};
  background: ${(props) =>
    props.selected
      ? props.theme.combobox.item.selectedBgColor
      : props.hover
      ? props.theme.combobox.item.hoverBgColor
      : props.disabled
      ? props.theme.combobox.item.disabledBgColor
      : props.theme.combobox.item.bgColor};
  color: ${(props) =>
    props.selected
      ? props.theme.combobox.item.selectedColor
      : props.hover
      ? props.theme.combobox.item.hoverColor
      : props.disabled
      ? props.theme.combobox.item.disabledColor
      : props.theme.combobox.item.color};
  &:hover {
    background: ${(props) => props.theme.combobox.item.hoverBgColor};
    color: ${(props) => props.theme.combobox.item.hoverColor};
  }
  -webkit-touch-callout: none; /* iOS Safari */
  -webkit-user-select: none; /* Safari */
  -khtml-user-select: none; /* Konqueror HTML */
  -moz-user-select: none; /* Old versions of Firefox */
  -ms-user-select: none; /* Internet Explorer/Edge */
  user-select: none; /* Non-prefixed version, currently
                                supported by Chrome, Edge, Opera and Firefox */
  ${(props) => (props.disabled ? "cursor: default;" : "cursor: pointer;")}
`

Item.defaultProps = {
  theme: {
    fontFamily: "tahoma, arial, helvetica, sans-serif",
    fontSize: {
      medium: "12px",
    },
    combobox: {
      item: {
        margin: 0,
        height: "18px",
        padding: "2px 6px",
        borderWidth: 0,
        borderStyle: "solid",
        borderColor: "#ccc",
        borderRadius: 0,
        bgColor: "#fff",
        color: "#000",
        hoverBgColor: "#eee",
        hoverColor: "#000",
        disabledBgColor: "#fff",
        disabledColor: "rgba(0, 0, 0, 0.33)",
        selectedBgColor: "#ccddf3",
        selectedColor: "#000",
      },
    },
  },
}

export const StyledInput = styled.input.attrs({ type: "search" })`
  box-sizing: border-box;
  min-width: 0px;
  width: 100%;
  margin: 0;
  padding: 0 0 0 3px;
  border: 0;
  background: transparent;
  color: currentColor;
  &:focus {
    outline: none;
  }
`

export interface StyledDivProps {
  focused?: boolean
  error?: boolean
  disabled?: boolean
}

export const FlexContainer = styled.div<StyledDivProps>`
  box-sizing: border-box;
  display: inline-flex;
  min-width: 150px;
  white-space: nowrap;
  align-items: center;
  justify-content: center;
  text-align: left;
  flex: 1;
  height: ${(props) => props.theme.combobox.height};
  border-radius: ${(props) => props.theme.combobox.borderRadius};
  margin: ${(props) => props.theme.combobox.margin};
  padding: ${(props) => props.theme.combobox.padding};
  font-size: ${(props) => props.theme.fontSize.medium};
  font-family: ${(props) => props.theme.fontFamily};
  background: ${(props) =>
    props.error
      ? props.theme.combobox.errorBgColor
      : props.focused
      ? props.theme.combobox.focusedBgColor
      : props.disabled
      ? props.theme.combobox.disabledBgColor
      : props.theme.combobox.bgColor};
  color: ${(props) =>
    props.error
      ? props.theme.combobox.errorColor
      : props.focused
      ? props.theme.combobox.focusedColor
      : props.disabled
      ? props.theme.combobox.disabledColor
      : props.theme.combobox.color};
  border: ${(props) =>
    `${props.theme.combobox.borderWidth} 
    ${props.theme.combobox.borderStyle} 
    ${
      props.error
        ? props.theme.combobox.errorBorderColor
        : props.focused
        ? props.theme.combobox.focusedBorderColor
        : props.disabled
        ? props.theme.combobox.disabledBorderColor
        : props.theme.combobox.borderColor
    }`};
  ${(props) =>
    props.focused
      ? `box-shadow: 0 0 5px ${
          props.error
            ? props.theme.combobox.errorBorderColor
            : props.theme.combobox.focusedBorderColor
        };`
      : ""}
  &:hover {
    ${(props) =>
      !props.focused && !props.error && !props.disabled
        ? `border: ${props.theme.combobox.borderWidth} 
      ${props.theme.combobox.borderStyle} 
      ${props.theme.combobox.hoverBorderColor};
      color: ${props.theme.combobox.hoverColor};`
        : ""}
    ${(props) =>
      !props.focused && !props.error && !props.disabled
        ? `background: ${props.theme.combobox.hoverBgColor};`
        : ""}
  }
`

FlexContainer.defaultProps = {
  theme: {
    fontSize: {
      medium: "12px",
    },
    fontFamily: "tahoma, arial, helvetica, sans-serif",
    combobox: {
      height: "24px",
      margin: 0,
      padding: 0,
      borderWidth: "1px",
      borderStyle: "solid",
      borderColor: "#ccc",
      borderRadius: 0,
      bgColor: "#fff",
      color: "black",
      focusedBorderColor: "#99bbe8",
      focusedColor: "black",
      focusedBgColor: "#fff",
      errorBorderColor: "red",
      errorColor: "black",
      errorBgColor: "#fff",
      disabledColor: "rgba(0, 0, 0, 0.75)",
      disabledBorderColor: "#ccc",
      disabledBgColor: "rgba(239, 239, 239, 0.5)",
      hoverColor: "black",
      hoverBorderColor: "#99bbe8",
      hoverBgColor: "#fff",
    },
  },
}

export const InputContainer = styled.div`
  height: 100%;
  display: flex;
  flex: 1;
`

export const ButtonContainer = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
`

export interface ComboBoxButtonProps
  extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  rotate: string
}

export const StyledButton = styled.button<ComboBoxButtonProps>`
  box-sizing: border-box;
  cursor: pointer;
  margin: 0;
  padding: 0;
  overflow: visible;
  font-size: 1.5rem;
  text-align: center;
  border-radius: 50%;
  height: ${(props) => props.theme.combobox.button.height};
  width: ${(props) => props.theme.combobox.button.width};
  background-color: transparent;
  border: 0;
  ${(props) => (props.rotate === "true" ? "transform: rotate(180deg);" : "")}
  color: ${(props) => props.theme.combobox.button.color};
  &:hover {
    background-color: ${(props) => props.theme.combobox.button.bgColor};
    transition: background-color 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
    opacity: 0.5;
  }
  &:focus {
    outline: none;
  }
  &:disabled {
    opacity: 0.7;
    pointer-events: none;
  }
`
StyledButton.defaultProps = {
  theme: {
    combobox: {
      button: {
        height: "22px",
        width: "22px",
        bgColor: "#f2f0f0",
        color: "#999",
      },
    },
  },
}

export const StyledSpan = styled.span`
  width: 100%;
  display: flex;
  align-items: center;
  justify-content: center;
  text-align: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
`

export const StyledSvg = styled.svg`
  margin: 0;
  padding: 0;
  fill: currentColor;
  width: 1em;
  height: 1em;
  display: inline-block;
  font-size: 1.5rem;
  transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  flex-shrink: 0;
`
