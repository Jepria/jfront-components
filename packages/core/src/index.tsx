export * from '@jfront-ui/app'
export * from '@jfront-ui/datepicker'
export * from '@jfront-ui/form'
//export * from '@jfront-ui/grid'
export * from '@jfront-ui/hooks'
//export * from '@jfront-ui/input'
export * from '@jfront-ui/layout'
//export * from '@jfront-ui/pagingbar'
//export * from '@jfront-ui/tabpanel'
//export * from '@jfront-ui/toolbar'
