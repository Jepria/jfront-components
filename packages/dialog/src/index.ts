export * from './SimpleDialog'
export * from './AlertDialog'
export * from './FormDialog'
export * from './ErrorDialog'
export * from './WarningDialog'