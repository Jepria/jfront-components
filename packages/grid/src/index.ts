export * from './common'
export * from './jepgrid'
