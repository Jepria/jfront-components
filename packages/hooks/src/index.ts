export * from "./useOnClickOutside"
export * from "./useRouteMapping"
export * from "./useTree/"
export * from "./useTree/types"
export * from "./useFilter"
