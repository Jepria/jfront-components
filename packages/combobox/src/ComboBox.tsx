import React, { useState, useRef, useEffect } from "react"
import styled from "styled-components"
import nextId from "react-id-generator"
import { OpenImage, LoadingImage, ExclamationImage } from "@jfront/ui-icons"
import { ComboBoxButton } from "./ComboBoxButton"

interface ItemProps {
  disabled?: boolean
  selected?: boolean
}

const Item = styled.div<ItemProps>`
  overflow: hidden;
  white-space: nowrap;
  height: 18px;
  padding: 2px 6px;
  cursor: pointer;
  font-family: tahoma, arial, helvetica, sans-serif;
  font-size: 12px;
  text-align: left;
  ${(props: ItemProps) =>
    props.selected ? "background: #ccddf3;" : "&:hover {background: #eee}"}
  -webkit-touch-callout: none; /* iOS Safari */
  -webkit-user-select: none; /* Safari */
  -khtml-user-select: none; /* Konqueror HTML */
  -moz-user-select: none; /* Old versions of Firefox */
  -ms-user-select: none; /* Internet Explorer/Edge */
  user-select: none; /* Non-prefixed version, currently
                                supported by Chrome, Edge, Opera and Firefox */
  ${(props) => (props.disabled ? "opacity: 0.33;" : "cursor: pointer;")}
`

const Popup = styled.div`
  -webkit-box-sizing: border-box;
  box-sizing: border-box;
  position: absolute;
  width: 98%;
  z-index: 5100;
  margin: 0;
  padding: 0;
  background: white;
  overflow: auto;
  -webkit-box-shadow: 2px 2px 1px -1px rgba(0, 0, 0, 0.34);
  -moz-box-shadow: 2px 2px 1px -1px rgba(0, 0, 0, 0.34);
  box-shadow: 2px 2px 1px -1px rgba(0, 0, 0, 0.34);
  &:focus {
    outline: none;
  }
`

const JepRiaButton = styled(OpenImage)`
  -webkit-box-sizing: border-box;
  box-sizing: border-box;
  display: inline-block;
  vertical-align: top;
  cursor: pointer;
  height: 24px;
  border-left: 1px solid #ccc;
  &:focus {
    outline: none;
  }
  &:hover {
    opacity: 0.5;
  }
`

interface StyledButtonProps {
  isOpen: boolean
}

const StyledInput = styled.input.attrs({ type: "search" })`
  display: -webkit-inline-box;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-flex: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  min-width: 0px;
  margin: 0;
  padding: 0;
  padding-left: 3px;
  font-family: tahoma, arial, helvetica, sans-serif;
  font-size: 12px;
  border: 0;
  &:focus {
    outline: none;
  }
`

interface StyledDivProps {
  focused?: boolean
  error?: boolean
}

const StyledDiv = styled.div<StyledDivProps>`
  display: -webkit-inline-box;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-flex: 1;
  -ms-flex-positive: 1;
  margin: 0;
  padding: 0;
  min-width: 150px;
  white-space: nowrap;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  justify-content: center;
  min-height: 24px;
  text-align: left;
  ${(props) =>
    props.focused
      ? `box-shadow: 0 0 5px rgba(81, 203, 238, 1);
         border: 1px solid rgba(81, 203, 238, 1);`
      : "border: 1px solid #ccc; border-top: 1px solid #999;"};
  ${(props) =>
    props.error
      ? props.focused
        ? `box-shadow: 0 0 5px red;
        border: 1px solid red;`
        : "border: 1px solid red;"
      : ""};
`

export interface ComboBoxItemProps {
  id?: string
  disabled?: boolean
  value: any
  selected?: boolean
  label: string
  onClick?: () => void
}

export const ComboBoxItem = React.forwardRef<HTMLDivElement, ComboBoxItemProps>(
  (
    { id, disabled, value, label, children, onClick, selected, ...rest },
    ref,
  ) => {
    return (
      <Item
        id={id}
        onClick={!disabled ? onClick : undefined}
        ref={ref}
        disabled={disabled}
        selected={selected}
        {...rest}
      >
        {React.Children.count(children) > 0 ? children : label}
      </Item>
    )
  },
)

export enum ComboBoxVariant {
  standard = "standard",
  jepria = "jepria",
}

export interface ComboBoxProps {
  id?: string
  children?: React.ReactNode[]
  name?: string
  className?: string
  disabled?: boolean
  style?: React.CSSProperties
  openOnFocus?: boolean
  clearOnBlur?: boolean
  options?: any[]
  variant?: string
  isLoading?: boolean
  error?: string
  initialValue?: any
  value?: any
  getOptionName?: (option: any) => string
  getOptionValue?: (option: any) => any
  renderItem?: (props: ComboBoxItemProps) => React.ReactNode
  onFocus?: (event: React.FocusEvent) => void
  onBlur?: (event: React.FocusEvent) => void
  onInputChange?: (event: React.ChangeEvent<HTMLInputElement>) => void
  onSelectionChange?: (name: string, value: any) => void
}

export const ComboBox = React.forwardRef<HTMLInputElement, ComboBoxProps>(
  (
    {
      id = nextId(),
      children,
      disabled,
      openOnFocus = true,
      clearOnBlur = false,
      className,
      name = "",
      style,
      options,
      variant = "standard",
      initialValue = null,
      value = null,
      isLoading,
      error,
      getOptionName,
      getOptionValue,
      renderItem,
      ...props
    },
    ref,
  ) => {
    if (children && options) {
      throw new Error("children and options props can't be applied together")
    }

    const [isOpen, setIsOpen] = useState(false)
    const [currentValue, setCurrentValue] = useState(
      initialValue != null ? initialValue : value != null ? value : undefined,
    )
    const [focused, setFocused] = useState(false)
    const [text, setText] = useState("")
    const outerDivRef = useRef<HTMLDivElement>(null)
    const popupRef = useRef<HTMLDivElement>(null)
    const currentValueRef = useRef<HTMLDivElement>(null)
    const defaultInputRef = useRef<HTMLInputElement>(null)
    const inputRef = ref || defaultInputRef

    useEffect(() => {
      if (value != null) setCurrentValue(value)
    }, [value])

    const onFocus = (e: React.FocusEvent) => {
      if (!isOpen && openOnFocus) {
        setIsOpen(true)
      }
      setFocused(true)
      if (props.onFocus) props.onFocus(e)
    }

    const onBlur = (e: React.FocusEvent) => {
      const { currentTarget } = e
      const relatedTarget = e.relatedTarget || document.activeElement
      if (isOpen) {
        if (relatedTarget === null) {
          setIsOpen(false)
          setFocused(false)
          return
        }
        let node = relatedTarget
        while (node !== null) {
          if (node === currentTarget) return
          node = (node as any).parentNode
        }
        setIsOpen(false)
      }
      setFocused(false)
      if (clearOnBlur && !currentValue) setText("")
      if (props.onBlur) props.onBlur(e)
    }

    const onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
      if (currentValue) {
        setCurrentValue(undefined)
        if (onChangeValue) {
          onChangeValue(name, undefined)
        }
      }
      setText(e.target.value)
      if (props.onInputChange) props.onInputChange(e)
    }

    const onChangeValue = (label: string, newValue: any) => {
      if (newValue !== currentValue) {
        if (initialValue != null || value == null) {
          setCurrentValue(newValue)
        }
        setText(label)
        setIsOpen(false)
        setFocused(false)
        if (props.onSelectionChange) props.onSelectionChange(name, newValue)
      }
    }

    const toggle = () => {
      setIsOpen(!isOpen)
    }

    useEffect(() => {
      if (isOpen && currentValue) {
        if (currentValueRef.current) {
          currentValueRef.current.scrollIntoView({
            behavior: "smooth",
            block: "start",
          })
        } else {
          popupRef.current?.scrollTo({
            top: document.getElementById(`${id}_${currentValue}`)?.offsetTop,
          })
        }
      }
      // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [isOpen, currentValue])

    useEffect(() => {
      if (initialValue != null || (value != null && text.length == 0)) {
        const defaultValue = initialValue || value
        if (options) {
          const option = options.find((option) => {
            return getOptionValue
              ? getOptionValue(option) === defaultValue
              : option.value === defaultValue
          })
          if (option) {
            setText(getOptionName ? getOptionName(option) : option.name)
          }
        } else if (children && React.Children.count(children) > 0) {
          const child = children.find((child) => {
            return (child as any)?.props.value === defaultValue
          })
          if (child) {
            setText((child as any)?.props.label)
          }
        }
      }
      // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [value])

    const getPopupTop = () => {
      const rect = outerDivRef.current?.getBoundingClientRect()
      const screenHeight = document.body.clientHeight
      if (rect && screenHeight - rect.bottom < 20 && rect.top > 22) {
        return undefined
      } else {
        return `${
          (outerDivRef.current?.offsetTop
            ? outerDivRef.current?.offsetTop
            : 0) +
          (outerDivRef.current?.offsetHeight
            ? outerDivRef.current?.offsetHeight
            : 0) +
          2
        }px`
      }
    }

    const getPopupBottom = () => {
      const rect = outerDivRef.current?.getBoundingClientRect()
      const screenHeight = document.body.clientHeight
      if (
        outerDivRef.current?.offsetTop &&
        rect &&
        screenHeight - rect.bottom < 20 &&
        rect.top > 22
      ) {
        return `${screenHeight - outerDivRef.current?.offsetTop + 4}px`
      } else {
        return undefined
      }
    }

    const getPopupLeft = () => {
      return `${
        outerDivRef.current?.offsetLeft ? outerDivRef.current?.offsetLeft : 0
      }px`
    }

    const getPopupWidth = () => {
      return `${
        outerDivRef.current?.offsetWidth ? outerDivRef.current?.offsetWidth : 0
      }px`
    }

    const getPopupHeight = () => {
      const rect = outerDivRef.current?.getBoundingClientRect()
      const screenHeight = document.body.clientHeight

      if (rect && screenHeight - rect.bottom < 20 && rect.top > 22) {
        if (rect && rect.top > 100) {
          return "100px"
        } else if (rect && rect.top > 20) {
          return `${rect.top - 2}px`
        } else {
          return "20px"
        }
      } else {
        if (rect && screenHeight - rect.bottom > 100) {
          return "100px"
        } else if (rect && screenHeight - rect.bottom > 20) {
          return `${document.body.clientHeight - rect.bottom - 2}px`
        } else {
          return "20px"
        }
      }
    }

    const renderChildren = () => {
      return React.Children.map(children, (item, index) => {
        if (!React.isValidElement(item)) {
          return null
        }
        const itemValue = item.props.value
        if (
          (!props.onInputChange && item.props.label.startsWith(text)) ||
          currentValue !== undefined ||
          props.onInputChange
        ) {
          return React.cloneElement(item, {
            id: `${id}_${itemValue}`,
            disabled: item.props.disabled || disabled,
            value: itemValue,
            selected: currentValue === itemValue,
            ref: currentValue === itemValue ? currentValueRef : null,
            onClick: () => onChangeValue(item.props.label, itemValue),
          })
        } else {
          return null
        }
      })
    }

    const renderOptions = () => {
      return options?.map((option) => {
        const itemValue = getOptionValue ? getOptionValue(option) : option.value
        const itemLabel = getOptionName ? getOptionName(option) : option.name
        const itemProps = {
          id: `${id}_${itemValue}`,
          key: itemValue,
          label: itemLabel,
          value: itemValue,
          selected: currentValue === itemValue,
          ref: currentValue === itemValue ? currentValueRef : null,
          onClick: () => onChangeValue(itemLabel, itemValue),
        }
        if (
          (!props.onInputChange && itemLabel.startsWith(text)) ||
          currentValue !== undefined ||
          props.onInputChange
        ) {
          if (renderItem) {
            return renderItem(itemProps)
          } else {
            return React.createElement(ComboBoxItem, itemProps)
          }
        } else {
          return null
        }
      })
    }

    const renderItems = () => {
      if (React.Children.count(children) > 0) {
        return renderChildren()
      } else if (options && options.length > 0) {
        return renderOptions()
      } else {
        return null
      }
    }

    return (
      <StyledDiv
        className={className}
        focused={focused}
        ref={outerDivRef}
        onBlur={onBlur}
        style={style}
        error={error !== undefined}
      >
        <StyledInput
          id={`${id}_input`}
          ref={inputRef}
          value={text}
          onFocus={onFocus}
          onChange={onChange}
          autoComplete="off"
        />
        {(variant === ComboBoxVariant.standard && (
          <ComboBoxButton
            id={`${id}_button`}
            tabIndex={-1}
            rotate={String(isOpen)}
            onClick={toggle}
            onFocus={() => setFocused(true)}
          />
        )) ||
          (variant === ComboBoxVariant.jepria && (
            <JepRiaButton
              id={`${id}_button`}
              tabIndex={-1}
              onClick={toggle}
              onFocus={() => setFocused(true)}
            />
          ))}
        {isOpen && (
          <Popup
            id={`${id}_popup`}
            ref={popupRef}
            tabIndex={0}
            style={{
              top: getPopupTop(),
              left: getPopupLeft(),
              bottom: getPopupBottom(),
              width: getPopupWidth(),
              maxHeight: getPopupHeight(),
            }}
          >
            {renderItems()}
          </Popup>
        )}
        {isLoading && <LoadingImage />}
        {error !== undefined && <ExclamationImage title={error} />}
      </StyledDiv>
    )
  },
)
