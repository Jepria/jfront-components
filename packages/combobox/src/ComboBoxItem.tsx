import React from "react"
import { Item } from "./styles"

export interface ComboBoxItemProps {
  id?: string
  disabled?: boolean
  value: any
  selected?: boolean
  hover?: boolean
  label: string
  onClick?: () => void
}

export const ComboBoxItem = React.forwardRef<HTMLDivElement, ComboBoxItemProps>(
  (
    { id, disabled, value, label, children, onClick, hover, selected, ...rest },
    ref,
  ) => {
    return (
      <Item
        id={id}
        onClick={!disabled ? onClick : undefined}
        ref={ref}
        disabled={disabled}
        hover={hover}
        selected={selected}
        {...rest}
      >
        {React.Children.count(children) > 0 ? children : label}
      </Item>
    )
  },
)
