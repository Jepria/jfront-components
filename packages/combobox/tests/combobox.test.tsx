import React from "react"
import { fireEvent, render, screen } from "@testing-library/react"
import { ComboBox, ComboBoxItem } from "../src"

test("Checking for the existence of an element ComboBox", () => {
  expect(render(<ComboBox />))
})

test("Checking text input in a field", () => {
  let idElementSelect = 0
  render(
    <ComboBox
      onSelectionChange={(name, value) => {
        idElementSelect = value
      }}
    >
      <ComboBoxItem data-testid="itemCheck" value={1} label="test1" />
      <ComboBoxItem value={2} label="test2" />
      <ComboBoxItem value={3} label="test14" />
    </ComboBox>,
  )

  fireEvent.click(screen.getByRole("button"))
  fireEvent.click(screen.getByTestId("itemCheck"))
  expect(idElementSelect).toEqual(1)
})
test("Matches snapshot ", () => {
  const { asFragment } = render(
    <ComboBox onSelectionChange={(name, value) => console.log(value)}>
      <ComboBoxItem value={1} label="test1" />
      <ComboBoxItem value={2} label="test2" />
      <ComboBoxItem value={3} label="test3" />
    </ComboBox>,
  )
  expect(asFragment()).toMatchSnapshot()
})

test("ComboBox isLoading renders correctly", () => {
  const tools = render(
    <>
      <ComboBox
        onSelectionChange={(name, value) => console.log(value)}
        isLoading
      >
        <ComboBoxItem value={1} label="test1" />
        <ComboBoxItem value={2} label="test2" />
      </ComboBox>
    </>,
  )
  expect(tools.asFragment()).toMatchSnapshot()
})
test("ComboBox isError renders correctly", () => {
  const tools = render(
    <>
      <ComboBox
        onSelectionChange={(name, value) => console.log(value)}
        style={{ width: "200px" }}
        error="error"
      >
        <ComboBoxItem value={1} label="test1" />
        <ComboBoxItem value={2} label="test2" />
      </ComboBox>
    </>,
  )
  expect(tools.asFragment()).toMatchSnapshot()
})
