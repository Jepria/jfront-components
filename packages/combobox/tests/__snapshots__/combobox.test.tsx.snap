// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ComboBox isError renders correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-AxmLO hQkLgg"
    style="width: 200px;"
  >
    <input
      autocomplete="off"
      class="sc-Axmtr brOIdg"
      id="id8_input"
      type="search"
      value=""
    />
    <button
      class="sc-AxiKw eRzEko"
      id="id8_button_button"
      rotate="false"
      tabindex="-1"
      type="button"
    >
      <span
        class="sc-AxhCb iEiKXF"
      >
        <svg
          class="sc-AxhUy eUkfkW"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </span>
    </button>
    <img
      class="sc-AxjAm khJDGy"
      src="data:image/gif;base64,R0lGODlhEAAQAIcAAED/QMJLMsNNNMVPNsNUPcxVPMZXQcddRs5dTNBZQcxiTM5kTtVgR9FjTNFhUNRkUthwWd5wWd1wXd18b919b+FdUuJfU+NhVOViVeZlVudnV+NmW+RnXOdsXehpWOpsWetuWulvX+F0XOVyWep2XexwW+5zXO92XfB4X/F6X+BzYO57Yel9aPN8YPN+YemCau2BaeyBbO6DbuOEdeaIeOqHfeuJfeqNe+yLf+ySf/KAZfSAYvWBYvWCY/SEZ/WEaveGbPaJa/iGbfONdvWOd/eMdfqOd/qUfN6hleWOg+eQhOeRheySiO2SieyYie2flfGTgvCXg/GXhPSZhPCfjvGdkOGkl+Onmu6glfihiPiiifijivOmkfSmkvGlmfGom/Gon/Kpn+uvovOvpvSsofWwo/ezoPezpPC0qPi0oPi2ofi1pfm5pvq8qPu9qfi+sfq/sv3DtfrZ0////wAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAACH5BAMAAAAALAAAAAAQABAAAAjBAAEIHEiwoME4R4D8IPLG4EAjMrqkScOFBQyHQqS42RIkiJY2OUYULBLFTY8ec+a4aMHmhoiBcGKY5LEjZQoUJ9ZIECNwCJcsO1amPGGiBBQnEASuUOOjxc2UJUB8CPElgUASZnTgNCFnzgcPGjp4KSDwBZUpRUFUGaMhAwYbSxYIRKPijFQPKTFcsBDGgZWBEWiUaduEiYUKYSgYKMhgBhkcHDbUADMhgMMGD5RgeZIEAQGHAq8oGCDgABLQqAkGBAA7"
      title="error"
    />
  </div>
</DocumentFragment>
`;

exports[`ComboBox isLoading renders correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-AxmLO kMckWu"
  >
    <input
      autocomplete="off"
      class="sc-Axmtr brOIdg"
      id="id7_input"
      type="search"
      value=""
    />
    <button
      class="sc-AxiKw eRzEko"
      id="id7_button_button"
      rotate="false"
      tabindex="-1"
      type="button"
    >
      <span
        class="sc-AxhCb iEiKXF"
      >
        <svg
          class="sc-AxhUy eUkfkW"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </span>
    </button>
    <img
      class="sc-AxirZ cxtyjb"
      src="data:image/gif;base64,R0lGODlhEAAQALMMAKqooJGOhp2bk7e1rZ2bkre1rJCPhqqon8PBudDOxXd1bISCef///wAAAAAAAAAAACH/C05FVFNDQVBFMi4wAwEAAAAh+QQFAAAMACwAAAAAEAAQAAAET5DJyYyhmAZ7sxQEs1nMsmACGJKmSaVEOLXnK1PuBADepCiMg/DQ+/2GRI8RKOxJfpTCIJNIYArS6aRajWYZCASDa41Ow+Fx2YMWOyfpTAQAIfkEBQAADAAsAAAAABAAEAAABE6QyckEoZgKe7MEQMUxhoEd6FFdQWlOqTq15SlT9VQM3rQsjMKO5/n9hANixgjc9SQ/CgKRUSgw0ynFapVmGYkEg3v1gsPibg8tfk7CnggAIfkEBQAADAAsAAAAABAAEAAABE2QycnOoZjaA/IsRWV1goCBoMiUJTW8A0XMBPZmM4Ug3hQEjN2uZygahDyP0RBMEpmTRCKzWGCkUkq1SsFOFQrG1tr9gsPc3jnco4A9EQAh+QQFAAAMACwAAAAAEAAQAAAETpDJyUqhmFqbJ0LMIA7McWDfF5LmAVApOLUvLFMmlSTdJAiM3a73+wl5HYKSEET2lBSFIhMIYKRSimFriGIZiwWD2/WCw+Jt7xxeU9qZCAAh+QQFAAAMACwAAAAAEAAQAAAETZDJyRCimFqbZ0rVxgwF9n3hSJbeSQ2rCWIkpSjddBzMfee7nQ/XCfJ+OQYAQFksMgQBxumkEKLSCfVpMDCugqyW2w18xZmuwZycdDsRACH5BAUAAAwALAAAAAAQABAAAARNkMnJUqKYWpunUtXGIAj2feFIlt5JrWybkdSydNNQMLaND7pC79YBFnY+HENHMRgyhwPGaQhQotGm00oQMLBSLYPQ9QIASrLAq5x0OxEAIfkEBQAADAAsAAAAABAAEAAABE2QycmUopham+da1cYkCfZ94UiW3kmtbJuRlGF0E4Iwto3rut6tA9wFAjiJjkIgZAYDTLNJgUIpgqyAcTgwCuACJssAdL3gpLmbpLAzEQA7"
    />
  </div>
</DocumentFragment>
`;

exports[`Matches snapshot  1`] = `
<DocumentFragment>
  <div
    class="sc-AxmLO kMckWu"
  >
    <input
      autocomplete="off"
      class="sc-Axmtr brOIdg"
      id="id6_input"
      type="search"
      value=""
    />
    <button
      class="sc-AxiKw eRzEko"
      id="id6_button_button"
      rotate="false"
      tabindex="-1"
      type="button"
    >
      <span
        class="sc-AxhCb iEiKXF"
      >
        <svg
          class="sc-AxhUy eUkfkW"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </span>
    </button>
  </div>
</DocumentFragment>
`;
